Refactor `ScrollingHitObjectContainer` and expose more useful methods

ekrctb opened the pull request:
The version of using local coordinate instead of screen space coordinate seems more useful to implement editor selection blueprint.
And the code is simplified by introducing a property `scrollingAxis`.
Not sure it is worth an `enum` because it is only used for private properties. If so, `Vector2` extension methods are desired.

ekrctb committed:
Refactor `ScrollingHitObjectContainer` and expose more useful methods

smoogipoo commented on osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs:
```diff
@@ -19,6 +19,11 @@ public class ScrollingHitObjectContainer : HitObjectContainer
+        /// <summary>
+        /// 0 for horizontal scroll, 1 for vertical scroll.
+        /// </summary>
+        private int scrollingAxis => direction.Value == ScrollingDirection.Left || direction.Value == ScrollingDirection.Right ? 0 : 1;
```
Maybe better to use `Axes.X`/`Axes.Y` instead? I'm not much of a fan of int magic values.

smoogipoo replied:
Oh I guess it's done this way for `scrollLength`...

bdach replied:
I think this could be less magicky if this was a method with a signature of `float getDimensionAlong(ScrollingDirection)` or similar. I *think* that should encompass all usages.

ekrctb replied:
`Axes` has `Axes.Both` so it is not the perfect fit.
And the reason I'm using `int` is the `Vector2` indexer.
A new `enum Axis { X, Y }` could be added, but perhaps this should be in osu-framework? The two `Axis` and `Axes` are confusing.

bdach replied:
I'm fine with a local enum, fwiw. Or even bringing back the old `getBreadth()` stuff, because if I'm honest, after reading this code a bit more to debug the failures this 0/1 stuff is a major readability regression.

bdach replied:
Actually, never mind. On checking `Direction` [already exists / is public API](https://github.com/ppy/osu-framework/blob/264f29abd5731dcc85c22c4b2d133b4efe41f210/osu.Framework/Graphics/Drawable.cs#L2865-L2869) in framework.

smoogipoo commented on osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs:
```diff
@@ -48,85 +53,65 @@ private void load()
 
         /// <summary>
-        /// Given a position in screen space, return the time within this column.
+        /// Given a position along the scrolling axis, return the time within this <see cref="HitObjectContainer"/>.
```
We generally don't mention local-space since screen-space is the exception.

bdach replied:
Well alright, I suppose IDE autocomplete should make this pretty clear as well.

smoogipoo committed:
Merge branch 'master' into refactor-scrolling-hoc-1

smoogipoo reviewed (dismissed):
I was going to reword the xmldocs a little bit, but I think that can be done in a separate PR if anyone wants to. It's quite hard to make the English work any better.

bdach commented:
Actual test failures here. Mania editor is also pretty severely broken, but only with downscroll.

bdach reviewed (changes requested):
as above

bdach commented on osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs:
```diff
@@ -262,24 +240,12 @@ private void updateLayoutRecursive(DrawableHitObject hitObject)
+            if (scrollingAxis == 0)
+                hitObject.X = position;
+            else
+                hitObject.Y = position;
```
This change is causing the test failures and needs to be reverted. The old code flipped the sign of `PositionAt()`, which is not the same as invoking `flipPositionIfRequired()`, as that does `DrawWidth - Y` or `DrawHeight - X`.

ekrctb commented:
Why there are two coordinate systems (different anchors) are used for `Down` and `Right` scroll in the first place?

bdach commented:
Well as far as I understand, `flipPositionIfRequired()` is doing a conversion of generalised "drawable space" (screen X and Y coordinates) to "scrolling space" (where position increases with time across the scrolling axis). So I believe what `updatePosition()` does is an inverse of that.

ekrctb committed:
Use `Direction` enum instead of `int`

The property is named `scrollingAxis` to distinguish from `direction`, which is of `ScrollingDirection` type (unfortunate name crash).

ekrctb committed:
Simplify `flipPositionIfRequired` using `scrollLength`

ekrctb commented:
I mean, there is countless number of code like:

```csharp
Anchor = Origin = e.NewValue == ScrollingDirection.Up ? Anchor.TopCentre : Anchor.BottomCentre;
```

in mania implementation.
Instead of using different `Anchor` for different scrolling directions, if `Anchor.TopLeft` was consistently used, `PositionAtTime` can be used to calculate the positions of the scrolling objects consistently, and mania implementation code is reduced.
I know this is a public API and it cannot be easily broken, but I'm wondering the current situation has some advantage over using the consistent anchor for all scrolling directions.

bdach commented:
Not sure why `TopLeft` in particular is the proposal. What about anchoring across the non-scrolling direction? And what would the top and left portions mean there exactly? Which axis is scrolling and which isn't?

I'm not necessarily opposed to doing something like

```csharp
Anchor Near { get; }
Anchor Far { get; }
```

where `near` is closer to the scroll target, but I'm not sure whether code using it will remain readable. The current state of affairs isn't *super* pretty but I think it holds up.

ekrctb commented on osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs:
```diff
@@ -48,85 +53,65 @@ private void load()
-        public double TimeAtScreenSpacePosition(Vector2 screenSpacePosition)
+        /// <param name="position">The position along the scrolling axis.</param>
+        /// <param name="referenceTime">The time the scrolling speed is used.</param>
+        public double TimeAtPosition(float position, double referenceTime)
```
This description of referenceTime is wrong. It should just be "current time".

ekrctb committed:
Fix `TimeAtPosition` doc comment

ekrctb committed:
Clarify multiple coordinate systems

- Fix wrong position is set for DHOs for down/right scrolling direction.

smoogipoo commented on osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs:
```diff
@@ -19,6 +19,16 @@ public class ScrollingHitObjectContainer : HitObjectContainer
+        private Direction scrollingAxis => direction.Value == ScrollingDirection.Left || direction.Value == ScrollingDirection.Right ? Direction.Horizontal : Direction.Vertical;
+
+        /// <summary>
+        /// Whether the scrolling direction is the positive-to-negative direction in the local coordinate.
```
Shouldn't this be negative-to-positive? So if the scroll direction is down, an object is scrolling from 0 -> Length.

ekrctb replied:
My thought was the axis is inverted when if time of an object is increased then the position of the object is decreased.
I agree it could be understood in the opposite way and confusing.
Not sure how to express this.

bdach replied:
I was also similarly confused by this xmldoc, so I've reworded it a touch and added an example to hopefully explain better. See if you think that's more clear than what was here before.

(Disabling auto-merge pending confirmation.)

smoogipoo reviewed (dismissed):
Looks good

smoogipoo committed:
Merge branch 'master' into refactor-scrolling-hoc-1

ekrctb commented on osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs:
```diff
@@ -262,24 +230,16 @@ private void updateLayoutRecursive(DrawableHitObject hitObject)
-                    break;
+            // The position returned from `PositionAtTime` is assuming the `TopLeft` anchor.
+            // A correction is needed because the hit objects are using a different anchor for each direction (e.g. `BottomCentre` for `Bottom` direction).
+            float anchorCorrection = axisInverted ? scrollLength : 0;
```
This cannot actually `hitObject.AnchorPosition` due to nested hit objects.

bdach replied:
Is that something that's blocking this pull from being merged, or some kind of a general note? Hard to tell what the intention of this comment is supposed to be.

ekrctb replied:
It is just a note.

bdach replied:
I was also similarly confused by this xmldoc, so I've reworded it a touch and added an example to hopefully explain better. See if you think that's more clear than what was here before.

(Disabling auto-merge pending confirmation.)

bdach committed:
Improve xmldoc

bdach reviewed (approved):
looks better

smoogipoo committed:
Merge branch 'master' into refactor-scrolling-hoc-1
